File: Java/card-game.java
public class Solution
{
    /**
     * @param n: The number of cards
     * @param totalProfit: The totalProfit
     * @param totalCost: The totalCost
     * @param a: The profit of cards
     * @param b: The cost of cards
     * @return: Return the number of legal plan
     */
    public int numOfPlan(int n, int totalProfit, int totalCost, int[] a, int[] b) {
        // Write your code here
        long mod = 1000000007;
        long [][][]dp;
        dp = new long [150][150][150];
        int i, j, k;
        for (i = 0; i <= 105; i++)
            for (j = 0; j <= 105; j++)
                for (k = 0; k <= 105; k++) {
                    dp[i][j][k] = 0;
                }
        dp[0][0][0] = 1;
        for (k = 0; k < n; k++)
            for (i = 0; i <= totalProfit + 1; i++)
                for (j = 0; j < totalCost; j++)  
                  if (dp[k][i][j] > 0) {
                        dp[k + 1][i][j] += dp[k][i][j];
                        dp[k + 1][i][j] %= mod;
                        if (j + b[k] < totalCost) {
                            dp[k + 1][Math.min(totalProfit + 1, i + a[k])][j + b[k]] += dp[k][i][j];
                            dp[k + 1][Math.min(totalProfit + 1, i + a[k])][j + b[k]] %= mod;
                        }
                    }
        long sum;
        sum = 0;
        for (j = 0; j < totalCost; j++) {
            sum = (sum + dp[n][totalProfit + 1][j]) % mod;
        }
        return (int)sum;
    }
}

File: C++/card-game.cpp
class Solution {
public:
    /**
     * @param n: The number of cards
     * @param totalProfit: The totalProfit
     * @param totalCost: The totalCost
     * @param a: The profit of cards
     * @param b: The cost of cards
     * @return: Return the number of legal plan
     */
    int numOfPlan(int n, int totalProfit, int totalCost, vector<int> &a, vector<int> &b) {
        // Write your code here
        int dp[110][110];
        int mod = 1e9 + 7;
        
        memset(dp, 0, sizeof(dp));
        dp[0][0] = 1;
        for (int i = 0; i < n; i++) {
            for (int p = totalProfit + 1; p >= 0; p--) {
                for (int c = totalCost + 1; c >= 0; c--) {
                    int now_p = min(totalProfit + 1, p + a[i]);
                    int now_c = min(totalCost + 1, c + b[i]);
                    dp[now_p][now_c] = (dp[now_p][now_c] + dp[p][c]) % mod;
                }   
            }
        }
        
        int ans = 0;
        for (int i = 0; i < totalCost; i++) {
            ans = (ans + dp[totalProfit + 1][i]) % mod;
        }
        return ans;
        
    }
};
